src/main/java/nl/b3p/tailormap/api/controller/MapController.java

```diff
@@ -149,48 +158,165 @@ private void getApplicationParams(@NotNull Application a, @NotNull MapResponse m
+        mapResponse.addBaseLayerTreeNodesItem(rootBackgroundNode);
+
+        List<Level> levels = levelRepository.findByLevelTree(a.getRoot().getId());
         List<StartLayer> startLayers = a.getStartLayers();
 
-        if (null != a.getStartLayers()) {
-            CoordinateReferenceSystem appCRS =
-                    new CoordinateReferenceSystem()
-                            // TODO use app projection, as TM model does not store
-                            //      app layer projection
-                            .code(ParseUtil.parseEpsgCode(a.getProjectionCode()))
-                            .definition(ParseUtil.parseProjDefintion(a.getProjectionCode()))
-                            .bounds(
-                                    ReferencingHelper.crsBoundsExtractor(
-                                            ParseUtil.parseEpsgCode(a.getProjectionCode())));
-
-            for (StartLayer l : startLayers) {
-                AppLayer appLayer =
-                        new AppLayer()
-                                .id(l.getApplicationLayer().getId())
-                                .displayName(l.getApplicationLayer().getLayerName())
-                                .url(l.getApplicationLayer().getService().getUrl())
-                                .serviceId(l.getApplicationLayer().getService().getId())
-                                .visible(l.isChecked())
-                                .crs(appCRS)
-                                .isBaseLayer(/*TODO fix hardcoded data, depends on HTM-247 */ true);
-
-                GeoService geoService = l.getApplicationLayer().getService();
-                Service s =
-                        new Service()
-                                .url(geoService.getUrl())
-                                .id(geoService.getId())
-                                .name(geoService.getName())
-                                .protocol(Service.ProtocolEnum.fromValue(geoService.getProtocol()))
-                                .capabilities(geoService.getCapabilitiesDoc());
-                if (geoService.getProtocol().equalsIgnoreCase(TileService.PROTOCOL)) {
-                    s.tilingProtocol(
-                            Service.TilingProtocolEnum.fromValue(
-                                    ((TileService) geoService).getTilingProtocol()));
+        if (startLayers == null) {
+            // Nothing to do.
+            return;
+        }
+
+        // TODO: sort list of startLayers?
+
+        List<Level> levelSet = new ArrayList<>();
+        for (Level level : levels) {
+            // Find all levels that contain a StartLayer.
+            boolean isInteresting = false;
+            for (ApplicationLayer layer : level.getLayers()) {
+                // TODO: O(n^2)-ish
+                for (StartLayer startLayer : startLayers) {
+                    if (startLayer.getApplicationLayer().getId() == layer.getId()) {
```

Comment:
```suggestion
                    if (startLayer.getApplicationLayer().getId().equals(layer.getId())) {
```

[BoxedPrimitiveEquality] Comparison using reference equality instead of value equality. Reference equality of boxed primitive types is usually not useful, as they are value objects, and it is bug-prone, as instances are cached for some values but not others.
    (see https://errorprone.info/bugpattern/BoxedPrimitiveEquality)